src/csync/csync_exclude.cpp

```diff
@@ -520,7 +548,62 @@ CSYNC_EXCLUDE_TYPE csync_excluded_traversal(CSYNC *ctx, const char *path, int fi
     return match;
 }
 
-CSYNC_EXCLUDE_TYPE csync_excluded_no_ctx(c_strlist_t *excludes, const char *path, int filetype) {
-  return _csync_excluded_common(excludes, path, filetype, true);
+CSYNC_EXCLUDE_TYPE ExcludedFiles::fullPatternMatch(const char *path, int filetype) const
+{
+    return _csync_excluded_common(_allExcludes, path, filetype, true);
+}
+
+auto ExcludedFiles::csyncTraversalMatchFun() const
+    -> std::function<CSYNC_EXCLUDE_TYPE(const char *path, int filetype)>
+{
+    return [this](const char *path, int filetype) { return this->traversalPatternMatch(path, filetype); };
+}
+
+/* Only for bnames (not paths) */
+static QString convertToBnameRegexpSyntax(QString exclude)
+{
+    QString s = QRegularExpression::escape(exclude).replace("\\*", ".*").replace("\\?", ".");
+    return s;
+}
+
+void ExcludedFiles::prepare()
+{
+    _nonRegexExcludes.clear();
+
+    // Start out with regexes that would match nothing
+    QString exclude_only = "a^";
+    QString exclude_and_remove = "a^";
+
+    for (auto exclude : _allExcludes) {
+        QString *builderToUse = &exclude_only;
+        if (exclude[0] == '\n')
+            continue; // empty line
+        if (exclude[0] == '\r')
+            continue; // empty line
+
+        /* If an exclude entry contains some fnmatch-ish characters, we use the C-style codepath without QRegularEpression */
+        if (strchr(exclude, '/') || strchr(exclude, '[') || strchr(exclude, '{') || strchr(exclude, '\\')) {
+            _nonRegexExcludes.append(exclude);
+            continue;
+        }
+
+        /* Those will attempt to use QRegularExpression */
+        if (exclude[0] == ']') {
+            exclude = exclude.mid(1);
+            builderToUse = &exclude_and_remove;
+        }
+        if (builderToUse->size() > 0) {
+            builderToUse->append("|");
+        }
+        builderToUse->append(convertToBnameRegexpSyntax(QString::fromUtf8(exclude)));
+    }
+
+    QString pattern = "^(" + exclude_only + ")$|^(" + exclude_and_remove + ")$";
+    _regex.setPattern(pattern);
+    QRegularExpression::PatternOptions patternOptions = QRegularExpression::OptimizeOnFirstUsageOption;
+    if (OCC::Utility::fsCasePreserving())
+        patternOptions |= QRegularExpression::CaseInsensitiveOption;
+    _regex.setPatternOptions(patternOptions);
+    _regex.optimize();
```

Comment:
Why pass the OptimizeOnFirstUsageOption flag if you optimize now?

Reply:
That was kept from the initial version. I agree that it can be removed.